autogen/agentchat/conversable_agent.py

```diff
@@ -933,8 +1077,9 @@ def generate_reply(
         By default, the following functions are checked in order:
         1. check_termination_and_human_reply
         2. generate_function_call_reply
-        3. generate_code_execution_reply
-        4. generate_oai_reply
+        3. generate_tool_calls_repl
```

Comment:
```suggestion
        3. generate_tool_calls_reply
```
When will generate_function_call_reply be useful? Is it only for old model and old client version, or also for new model and new client version? The answer to this question will decide whether we want to keep the two reply methods separate.

Reply:
Yeah generate_function_call_reply backwards compatibility at this point.

Definitely another way to approach this would be to just make all of the function based interface just auto translate to tools. Top of the head I don't think there is any functional difference at this point, tools is a superset of functions that doesn't actually add any features yet (well multi-call, but that isn't really related to the create definition). So far have assumed deprecation or combining the concerns would be a seperate PR.

Reply:
Thanks for the clarification. If the plan is to deprecate `generate_function_call_reply`, could you add a note in the docstr about what model and what client version require this reply method to be available? I'm not clear about the conditions.

Reply:
Yeah, I can add

```
        Deprecated as of [OpenAI API v1.1.0](https://github.com/openai/openai-python/releases/tag/v1.1.0)
        See https://platform.openai.com/docs/api-reference/chat/create#chat-create-function_call
```

on the related methods.

I don't know if there are any particular OpenAI models that support function_call but not tool_calls or the other way around. I'd imagine microsoft and projects policies would go into the deprecation PR, whatever those may be. In the meantime both are equally supported, so far I'm not finding a deadline on OpenAI removing function_call and seems like all current models support both methods.

Reply:
I think this notice looks good. Perhaps add a bit more context:

```
`functions` keyword argument was replaced by `tools` as of [OpenAI API v1.1.0](https://github.com/openai/openai-python/releases/tag/v1.1.0)
See https://platform.openai.com/docs/api-reference/chat/create#chat-create-function_call
```